### PR TITLE
Autologging performance improvements for XGBoost / LightGBM

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?

Autologging performance improvements for XGBoost / LightGBM

## How is this patch tested?

Existing unit tests for correctness

## Release Notes

Improved autologging performance for XGBoost and LightGBM.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/xgboost/test_xgboost_autolog.py [hunk 1]
```diff
@@ -249,49 +249,6 @@ def test_xgb_autolog_logs_metrics_with_early_stopping(bst_params, dtrain):
             assert metric_history == evals_result[eval_name][metric_name] + [best_metrics]
 
 
-@pytest.mark.large
-def test_xgb_autolog_batch_metrics_logger_logs_expected_metrics(bst_params, dtrain):
```
Comment (dbczumar) [inline tests/xgboost/test_xgboost_autolog.py:253]: This test case is unnecessary - we have several other test cases that ensure correct metrics are logged as a result of training. We now have queued / batched logging for metrics, params, & tags in addition to the preexisting `BatchMetricsLogger` for per-epoch metric logging from callbacks; having a test case asserting that a particular framework is used for particular metric values is not particularly valuable.

File: mlflow/lightgbm.py [hunk 2]
```diff
@@ -462,13 +471,22 @@ def log_feature_importance_plot(features, importance, importance_type):
...
+                    run_id=mlflow.active_run().info.run_id,
+                    metrics={
+                        "stopped_iteration": extra_step,
+                        # best_iteration is set even if training does not stop early.
+                        "best_iteration": model.best_iteration,
+                    },
+                )
                 # iteration starts from 1 in LightGBM.
-                results = eval_results[model.best_iteration - 1]
-                metrics_logger.record_metrics(results, step=extra_step)
```
Comment (dbczumar) [inline mlflow/lightgbm.py:471]: `BatchMetricsLogger` is tailored to per-epoch metric logging use cases from callbacks; it may make multiple API calls during this segment. The new `MlflowAutologgingQueueingClient` is a better choice here; it will only make a single API call to LogBatch.

File: mlflow/xgboost.py [hunk 3]
```diff
@@ -514,9 +522,19 @@ def log_feature_importance_plot(features, importance, importance_type):
             )
             if early_stopping:
                 extra_step = len(eval_results)
-                metrics_logger.record_metrics({"stopped_iteration": extra_step - 1})
-                metrics_logger.record_metrics({"best_iteration": model.best_iteration})
-                metrics_logger.record_metrics(eval_results[model.best_iteration], extra_step)
```
Comment (dbczumar) [inline mlflow/xgboost.py:519]: `BatchMetricsLogger` is tailored to per-epoch metric logging use cases from callbacks; it may make multiple logging API calls during this segment. The new `MlflowAutologgingQueueingClient` is a better choice here; it will only make a single API call to LogBatch.

File: mlflow/lightgbm.py (hunk 2 above)
Comment (jinzhang21) [reply mlflow/lightgbm.py:471]: Nice.

File: mlflow/xgboost.py [hunk 4]
```diff
@@ -461,15 +461,16 @@ def log_feature_importance_plot(features, importance, importance_type):
...
+                mlflow.log_artifact(filepath)
             finally:
                 plt.close(fig)
                 shutil.rmtree(tmpdir)
 
-        # logging booster params separately via mlflow.log_params to extract key/value pairs
-        # and make it easier to compare them across runs.
-        params = args[0] if len(args) > 0 else kwargs["params"]
-        try_mlflow_log(mlflow.log_params, params)
+        autologging_client = MlflowAutologgingQueueingClient()
```
Comment (jinzhang21) [inline mlflow/xgboost.py:469]: Not specific to this PR. But the train() function is getting really long and difficult to read. It might be good to factor some code out (e.g. move scoped functions to the top, and use with statement for both autologging_client) in a follow up cleanup PR.

**Review by jinzhang21 — APPROVED** [review]

File: mlflow/xgboost.py (hunk 4 above)
Comment (dbczumar) [reply mlflow/xgboost.py:469]: Agreed! I'll file a ticket to track cleaning this up
